### PR TITLE
[SILOpt] Add ZeroInitializer to isValidStaticInitializerInst

### PR DESCRIPTION
**Pull request by davezarzycki**

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Ping. Hi @gottesmm and @eeckstein – This seems like a simple QoI fix, right? Am I missing anything?

**Review by eeckstein — CHANGES_REQUESTED** [review]
This builtin needs to be supported in IRGen's emitConstantValue, too.
Can you add an "-emit-ir" test as well (you can do it in the same test file)?

**Comment by davezarzycki** [comment]
Hi @eeckstein – I feel like I'm missing something. Why does `emitConstantValue` need to special case `BuiltinValueKind::ZeroInitializer` when the code gen for this builtin already emits a constant? I ask because I'm not observing any behavioral differences with your suggestion.

**Comment by davezarzycki** [comment]
I did some sanity checks. The code you requested is definitely being called but it's not making an observable difference in the IR with the simple test I provided. Do I need some non-trivial Swift/SIL to observe the change you requested?

**Comment by eeckstein** [comment]
Hm, don't you hit the
```
llvm_unreachable("unsupported builtin for constant expression");
```
?

**Comment by rjmccall** [comment]
I think the key question is whether this is actually emitted as a constant initializer.  I don't know if changing this one SIL predicate is sufficient for that.

**Comment by davezarzycki** [comment]
My mistake. I wasn't doing a proper A versus B comparison. Patch updated.

@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Updated CHECK to use a regular expression to account for macOS versus Linux differences.

@swift-ci please smoke test

File: lib/IRGen/GenConstant.cpp [hunk 1]
```diff
@@ -56,6 +56,24 @@ llvm::Constant *irgen::emitConstantInt(IRGenModule &IGM,
...
+llvm::Constant *irgen::emitConstantZero(IRGenModule &IGM, BuiltinInst *BI) {
+  assert(IGM.getSILModule().getBuiltinInfo(BI->getName()).ID ==
+         BuiltinValueKind::ZeroInitializer);
+  BuiltinIntegerWidth width =
+      BI->getType().castTo<AnyBuiltinIntegerType>()->getWidth();
+
+  APInt value(width.getLeastWidth(), 0);
+
+  // Handle arbitrary-precision integers.
+  if (width.isArbitraryWidth()) {
```
Comment (eeckstein) [inline lib/IRGen/GenConstant.cpp:68]: Can you add a test for this case?

File: lib/IRGen/GenConstant.cpp [hunk 2]
```diff
@@ -94,9 +112,11 @@ static llvm::Constant *emitConstantValue(IRGenModule &IGM, SILValue operand) {
...
   } else if (auto *BI = dyn_cast<BuiltinInst>(operand)) {
     switch (IGM.getSILModule().getBuiltinInfo(BI->getName()).ID) {
-      case BuiltinValueKind::PtrToInt: {
-        llvm::Constant *ptr = emitConstantValue(IGM, BI->getArguments()[0]);
-        return llvm::ConstantExpr::getPtrToInt(ptr, IGM.IntPtrTy);
+    case BuiltinValueKind::ZeroInitializer:
+      return emitConstantZero(IGM, BI);
+    case BuiltinValueKind::PtrToInt: {
+      llvm::Constant *ptr = emitConstantValue(IGM, BI->getArguments()[0]);
+      return llvm::ConstantExpr::getPtrToInt(ptr, IGM.IntPtrTy);
```
Comment (eeckstein) [inline lib/IRGen/GenConstant.cpp:119]: Can you fix the indentation here?

File: lib/IRGen/GenConstant.cpp (hunk 1 above)
Comment (davezarzycki) [reply lib/IRGen/GenConstant.cpp:68]: Sure, is there a good arbitrary width constant test I can copy from?

File: lib/IRGen/GenConstant.cpp (hunk 2 above)
Comment (davezarzycki) [reply lib/IRGen/GenConstant.cpp:119]: Hmm... I just ran the code through `git clang-format` and this was the sub-optimal result. I can either re‑indent all of the case statements to match clang style, or have this change match the rest of the case statements. What would you prefer?

**Comment by davezarzycki** [comment]
I dropped support for arbitrary width integers for two reasons: 1) using `Builtin.zeroInitializer()` with `Builtin.IntLiteral` type would be strange and 2) making that work would require more builtins to be handled in SIL and IRGen's constant handling logic.

I also overrode `git clang-format` and made the new case statement match the neighboring case statements.

@swift-ci please smoke test

File: lib/SIL/IR/SILGlobalVariable.cpp [hunk 3]
```diff
@@ -109,6 +109,8 @@ bool SILGlobalVariable::isValidStaticInitializerInst(const SILInstruction *I,
     case SILInstructionKind::BuiltinInst: {
       auto *bi = cast<BuiltinInst>(I);
       switch (M.getBuiltinInfo(bi->getName()).ID) {
+        case BuiltinValueKind::ZeroInitializer:
```
Comment (eeckstein) [inline lib/SIL/IR/SILGlobalVariable.cpp:112]: Can you add a check here, which returns false if it's an arbitrary width integer?

File: lib/IRGen/GenConstant.cpp (hunk 2 above)
Comment (eeckstein) [reply lib/IRGen/GenConstant.cpp]: I don't mind

**Comment by davezarzycki** [comment]
Updated to have `isValidStaticInitializerInst` return false if the zero initializer has arbitrary width.

@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Replaced uses of `AnyBuiltinIntegerType` with `BuiltinIntegerType` to avoid arbitrary width problems.

@swift-ci please smoke test

**Comment by davezarzycki** [comment]
Added float and vector support. What's left before a LGTM?

@swift-ci please smoke test

**Comment by rjmccall** [comment]
Not supporting `BuiltinIntegerLiteralType` is appropriate since IIUC we want this builtin to guarantee a zero-bit representation, which is not a valid inhabitant of that type (there always has to be at least one word of data, to avoid that being a special case).

**Review by eeckstein — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**